### PR TITLE
[DEVOPS-1316] Fix deadlocks and improves general performance

### PR DESCRIPTION
**Pull request by taksan**

### CODE REVIEW

File: core/src/main/java/objective/taskboard/jira/ProjectService.java [hunk 1]
```diff
@@ -87,10 +96,9 @@ private void updateProjectNameFromJira() {
     }
     
     public List<ProjectFilterConfiguration> getTaskboardProjects() {
-        return projectRepository.findAllByOrderByJiraProjectKeyAscProjectKeyAsc();
+        return cacheManager.getCache(TASKBOARD_PROJECT_LIST).get(SimpleKey.EMPTY, ()-> projectRepository.findAllByOrderByJiraProjectKeyAscProjectKeyAsc());
```
Comment (objective-solutions-user) [inline core/src/main/java/objective/taskboard/jira/ProjectService.java]: MINOR Replace this lambda with a method reference. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1612)

File: core/src/main/java/objective/taskboard/utils/persistence/PersistenceListenerManager.java [hunk 2]
```diff
@@ -1,33 +1,113 @@
...
+import org.springframework.aop.framework.Advised;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Component;
 
+@SuppressWarnings("serial")
 @Component
-public class PersistenceListenerManager {
+public class PersistenceListenerManager implements PostInsertEventListener, PostUpdateEventListener{
     @PersistenceUnit
     private EntityManagerFactory emf;
```
Comment (objective-solutions-user) [inline core/src/main/java/objective/taskboard/utils/persistence/PersistenceListenerManager.java]: CRITICAL Make "emf" transient or serializable. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1948)

File: core/src/main/java/objective/taskboard/utils/persistence/PersistenceListenerManager.java [hunk 3]
```diff
@@ -1,33 +1,113 @@
...
+
+    private void addPostInsertListener(EventDelegate eventDelegate) {
+        if (eventDelegate.method.getAnnotation(PostPersist.class) == null) return;
+        postUpdate.add(eventDelegate);
+    }
+
+    private void triggerEvent(EventDelegate delegate, Object entity) {
+        if (!delegate.targetEntityClass.isAssignableFrom(entity.getClass())) return;
+        try {
+            delegate.method.invoke(delegate.listener, new Object[] { entity });
```
Comment (objective-solutions-user) [inline core/src/main/java/objective/taskboard/utils/persistence/PersistenceListenerManager.java]: MINOR Remove this array creation and simply pass the elements. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS3878)

File: core/src/main/java/objective/taskboard/utils/persistence/PersistenceListenerManager.java [hunk 4]
```diff
@@ -1,33 +1,113 @@
...
 import org.springframework.stereotype.Component;
 
+@SuppressWarnings("serial")
 @Component
-public class PersistenceListenerManager {
+public class PersistenceListenerManager implements PostInsertEventListener, PostUpdateEventListener{
     @PersistenceUnit
     private EntityManagerFactory emf;
     
+    private List<PostPersistenceListener> listeners;
```
Comment (objective-solutions-user) [inline core/src/main/java/objective/taskboard/utils/persistence/PersistenceListenerManager.java]: CRITICAL Make "listeners" transient or serializable. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1948)

File: core/src/main/java/objective/taskboard/utils/persistence/PersistenceListenerManager.java [hunk 5]
```diff
@@ -1,33 +1,113 @@
...
+    private List<EventDelegate> postUpdate = new LinkedList<>();
```
Comment (objective-solutions-user) [inline core/src/main/java/objective/taskboard/utils/persistence/PersistenceListenerManager.java]: CRITICAL Make "postUpdate" transient or serializable. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1948)

File: core/src/main/java/objective/taskboard/utils/persistence/PersistenceListenerManager.java [hunk 6]
```diff
@@ -1,33 +1,113 @@
...
+    private List<EventDelegate> postInsert = new LinkedList<>();
```
Comment (objective-solutions-user) [inline core/src/main/java/objective/taskboard/utils/persistence/PersistenceListenerManager.java]: CRITICAL Make "postInsert" transient or serializable. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1948)

File: core/src/main/java/objective/taskboard/utils/persistence/PersistenceListenerManager.java [hunk 7]
```diff
@@ -1,33 +1,113 @@
...
+    private EventListenerRegistry registry;
```
Comment (objective-solutions-user) [inline core/src/main/java/objective/taskboard/utils/persistence/PersistenceListenerManager.java]: MINOR Remove the "registry" field and declare it as a local variable in the relevant methods. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS1450)

File: strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioChartService.java [hunk 8]
```diff
@@ -40,12 +41,11 @@
...
+import objective.taskboard.utils.persistence.PostPersistenceListener;
 
-@SuppressWarnings("serial")
 @CacheConfig(cacheNames = SCENARIO_CHART_DATA)
 @Service
-public class ScenarioChartService implements PostUpdateTaskboardPersistenceListener {
+public class ScenarioChartService implements PostPersistenceListener {
     private static final Logger log = getLogger(ScenarioChartService.class);
 
     private transient ScopeScenarioService scopeScenarioService;
```
Comment (objective-solutions-user) [inline strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioChartService.java]: MINOR Remove the "transient" modifier from this field. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2065)

**Comment by objective-solutions-user** [comment]
SonarQube analysis reported 16 issues
* ![CRITICAL][CRITICAL] 4 critical
* ![MINOR][MINOR] 12 minor

Watch the comments in this conversation to review them.

#### 8 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [ScenarioChartService.java#L52](https://github.com/objective-solutions/taskboard/blob/58e3679aee9ec61b1d2d86d6925d0d223eee278f/strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioChartService.java#L52): Remove the "transient" modifier from this field. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2065)
1. ![MINOR][MINOR] [ScenarioChartService.java#L53](https://github.com/objective-solutions/taskboard/blob/58e3679aee9ec61b1d2d86d6925d0d223eee278f/strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioChartService.java#L53): Remove the "transient" modifier from this field. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2065)
1. ![MINOR][MINOR] [ScenarioChartService.java#L54](https://github.com/objective-solutions/taskboard/blob/58e3679aee9ec61b1d2d86d6925d0d223eee278f/strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioChartService.java#L54): Remove the "transient" modifier from this field. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2065)
1. ![MINOR][MINOR] [ScenarioChartService.java#L55](https://github.com/objective-solutions/taskboard/blob/58e3679aee9ec61b1d2d86d6925d0d223eee278f/strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioChartService.java#L55): Remove the "transient" modifier from this field. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2065)
1. ![MINOR][MINOR] [ScenarioChartService.java#L56](https://github.com/objective-solutions/taskboard/blob/58e3679aee9ec61b1d2d86d6925d0d223eee278f/strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioChartService.java#L56): Remove the "transient" modifier from this field. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2065)
1. ![MINOR][MINOR] [ScenarioChartService.java#L57](https://github.com/objective-solutions/taskboard/blob/58e3679aee9ec61b1d2d86d6925d0d223eee278f/strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioChartService.java#L57): Remove the "transient" modifier from this field. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2065)
1. ![MINOR][MINOR] [ScenarioChartService.java#L58](https://github.com/objective-solutions/taskboard/blob/58e3679aee9ec61b1d2d86d6925d0d223eee278f/strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioChartService.java#L58): Remove the "transient" modifier from this field. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2065)
1. ![MINOR][MINOR] [ScenarioChartService.java#L59](https://github.com/objective-solutions/taskboard/blob/58e3679aee9ec61b1d2d86d6925d0d223eee278f/strategical-dashboard/src/main/java/objective/taskboard/monitor/ScenarioChartService.java#L59): Remove the "transient" modifier from this field. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2065)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: core/src/main/java/objective/taskboard/project/config/ProjectClusterHistoryService.java [hunk 9]
```diff
@@ -21,12 +22,12 @@
...
 import objective.taskboard.followup.cluster.SizingClusterItemRepository;
+import objective.taskboard.utils.persistence.PostPersistenceListener;
+
 
-@SuppressWarnings("serial")
 @Service
-public class ProjectClusterHistoryService implements PostUpdateTaskboardPersistenceListener {
+public class ProjectClusterHistoryService implements PostPersistenceListener {
 
     private final transient AuditService auditService;
```
Comment (objective-solutions-user) [inline core/src/main/java/objective/taskboard/project/config/ProjectClusterHistoryService.java]: MINOR Remove the "transient" modifier from this field. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2065)

File: core/src/main/java/objective/taskboard/project/config/ProjectClusterHistoryService.java [hunk 10]
```diff
@@ -21,12 +22,12 @@
...
     private final transient BaseClusterService baseClusterService;
```
Comment (objective-solutions-user) [inline core/src/main/java/objective/taskboard/project/config/ProjectClusterHistoryService.java]: MINOR Remove the "transient" modifier from this field. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2065)

**Comment by objective-solutions-user** [comment]
SonarQube analysis reported 4 issues
* ![MINOR][MINOR] 4 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [ProjectClusterHistoryService.java#L34](https://github.com/objective-solutions/taskboard/blob/9175994b9a2ae30811244ee0cd46a483d5de61c5/core/src/main/java/objective/taskboard/project/config/ProjectClusterHistoryService.java#L34): Remove the "transient" modifier from this field. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2065)
1. ![MINOR][MINOR] [ProjectClusterHistoryService.java#L35](https://github.com/objective-solutions/taskboard/blob/9175994b9a2ae30811244ee0cd46a483d5de61c5/core/src/main/java/objective/taskboard/project/config/ProjectClusterHistoryService.java#L35): Remove the "transient" modifier from this field. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2065)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: core/src/main/java/objective/taskboard/issueBuffer/IssueBufferServiceSearchVisitor.java [hunk 11]
```diff
@@ -1,23 +1,33 @@
...
 
+import java.util.Deque;
+import java.util.concurrent.ConcurrentLinkedDeque;
+
+import objective.taskboard.data.IssueScratch;
 import objective.taskboard.domain.converter.JiraIssueToIssueConverter;
 import objective.taskboard.domain.converter.ParentProvider;
 import objective.taskboard.jira.client.JiraIssueDto;
 
 public class IssueBufferServiceSearchVisitor extends BaseSearchVisitor {
```
Comment (herbert-venancio) [inline core/src/main/java/objective/taskboard/issueBuffer/IssueBufferServiceSearchVisitor.java]: is `BaseSearchVisitor` still necessary?

**Comment by objective-solutions-user** [comment]
SonarQube analysis reported 4 issues
* ![MINOR][MINOR] 4 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [ProjectClusterHistoryService.java#L34](https://github.com/objective-solutions/taskboard/blob/0b00caa61036ac7263d856744d554df433743a5d/core/src/main/java/objective/taskboard/project/config/ProjectClusterHistoryService.java#L34): Remove the "transient" modifier from this field. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2065)
1. ![MINOR][MINOR] [ProjectClusterHistoryService.java#L35](https://github.com/objective-solutions/taskboard/blob/0b00caa61036ac7263d856744d554df433743a5d/core/src/main/java/objective/taskboard/project/config/ProjectClusterHistoryService.java#L35): Remove the "transient" modifier from this field. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2065)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by objective-solutions-user** [comment]
SonarQube analysis reported 4 issues
* ![MINOR][MINOR] 4 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [ProjectClusterHistoryService.java#L34](https://github.com/objective-solutions/taskboard/blob/eb77bb7bff8b9f395c3d6401d1447fec836c204a/core/src/main/java/objective/taskboard/project/config/ProjectClusterHistoryService.java#L34): Remove the "transient" modifier from this field. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2065)
1. ![MINOR][MINOR] [ProjectClusterHistoryService.java#L35](https://github.com/objective-solutions/taskboard/blob/eb77bb7bff8b9f395c3d6401d1447fec836c204a/core/src/main/java/objective/taskboard/project/config/ProjectClusterHistoryService.java#L35): Remove the "transient" modifier from this field. [rule](http://10.44.1.250:9000/coding_rules#rule_key=squid%3AS2065)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**